### ISSUE TITLE
[Bringup]- Bring up video+text path on-device at TP=8

### ISSUE THREAD

**Issue by saiarthiraguram** [tenstorrent/tt-xla#5266]
**Description:**

Arch: n300-llmbox (8× Wormhole_b0) · Mesh: (1, 8) Megatron 1D TP · Dtype: bf16 · real weights.
Harness: `mm_harness.py --modality video --num-frames N` · CPU sanity: `mm_av_cpu_sanity.py` → `mm_av_cpu_sanity.json`, `mm_golden_video.pt`.
Inputs built by the loader: `ModelLoader.load_inputs(include_video=True, num_frames=N)` → `load_video_inputs`.

**Status:** CPU sanity ✅ and on-device TP=8 ✅ both done at a reduced 4-frame clip, **last-token PCC 0.9969** vs the CPU golden (§3); through the runner the **full-tensor PCC is 0.9249** (§5). The full 32-frame clip is **activation-bound (OOM)** on a single chip — see §4. Video tower sharding is the follow-up.

## 1. Gemma4 video interface (what feeds the video path)

Gemma4Unified uses the same encoder-free patch embedder for video frames as for images, applied per frame: raw merged pixel patches → embedder → factorized 2D positional embedding → projected into the 3840-d text space and `masked_scatter`-ed into the token stream at video-token positions.

`Gemma4UnifiedProcessor` turns `"<|video|>" + text` + a stack of frames into:

| input | shape | dtype | role |
|-------|-------|-------|------|
| input_ids | (1, P) | int64 | text tokens with video-token (258884) span |
| attention_mask | (1, P) | int64 | standard mask |
| mm_token_type_ids | (1, P) | int64 | 0=text, 1=video |
| pixel_values_videos | (1, T, 70, 6912) | float32 | merged model patches, per frame (T frames × 70 patches) |
| video_position_ids | (1, T, 70, 2) | int64 | per-patch (x,y) per frame |

Each frame contributes ~64 video tokens, so the sequence is ~`64 × num_frames`. No video-decode backend (av/decord) is available, so the loader builds a static clip by replicating the sample image into `num_frames` frames and passes `do_sample_frames=False` so the frame count is honored exactly.

## 2. CPU sanity (reference) — ✅ PASS

`mm_av_cpu_sanity.py` ran one video+text forward on CPU (bf16, full 32-frame clip) with prompt `"<|video|>Describe the video."`:

| field | value |
|-------|-------|
| processor | Gemma4UnifiedProcessor |
| seq len P | 2308 (2048 video tokens + text/control) |
| pixel_values_videos | (1, 32, 70, 6912) |
| logits | (1, 2308, 262144), all finite ✅ |
| next-token argmax | 236743 → `' '` |
| top-5 | `' '`, `'0'`, `'\n'`, `'  '`, `<eos>` |

The full-clip forward is finite on CPU. For the on-device PCC, a matching 4-frame golden (seq 292, next 236743, finite) is saved to `mm_golden_video.pt`.

## 3. On-device (TP=8) — ✅ VERIFIED (4-frame clip)

Run: `mm_harness.py --modality video --use-loader --num-frames 4 --min-new-tokens 10` on n300-llmbox, mesh (1, 8), bf16, real weights, real `pixel_values_videos`. Setup: prompt+video P=292 (252 video tokens), 240 text-decoder weights sharded, video embedder weights replicated, `use_cache=False`. Result in `mm_results_video.json`, log `logs/mm_harness_video_1658.log`.

| check | result |
|-------|--------|
| Video-path forward compiles on tt backend | ✅ (per-frame patch embedder + masked_scatter lower fine) |
| Logits finite on-device | ✅ |
| PCC(device prefill last-token, CPU golden) | **0.9969** |
| Next-token argmax vs CPU | match — both 236743 (`' '`) |
| Fit at seq 320 | ✅ fits |
| Prefill TTFT (incl. one-time compile) | 187.9 s |
| Decode | ~3.42 s/step (full (1,320) forward, no KV cache) |

**Results — caption (eos-suppressed, `--min-new-tokens 10`):**

> `00:00  00`

## 4. Full 32-frame clip — ❌ OOM (activation-bound)

Run: `mm_harness.py --modality video --num-frames 32` (P=2308, buffer auto-raised to 2328). Log `logs/mm_harness_video_1647.log`.

```
TT_FATAL: Out of Memory: Not enough space to allocate 15854469120 B DRAM buffer
across 12 banks, where each bank needs to store 1321205760 B, but bank size is
1071821792 B (allocated: 318521568 B, free: 753300224 B)
```

At 2048 video tokens (32 frames) the embedding/projection op (`tilize_with_val_padding` → `ttnn::embedding`) requires a ~15.8 GB DRAM activation buffer — exceeds single-chip DRAM with the video tower replicated. The 4-frame clip (252 tokens) is the tractable first bring-up; the full clip needs the video tower sharded across the mesh — tracked here as follow-up.

## 5. Through the tt-xla runner (TP=8) — full-tensor PCC

Node `gemma4/pytorch-12B-video-tensor_parallel-inference` (variant `12B-video`, `VIDEO_NUM_FRAMES=4`), run via `tests/runner/test_models.py` on n300-llmbox.

| check | result |
|-------|--------|
| Runs end-to-end on tt backend | ✅ |
| Full-output-tensor PCC vs CPU | **0.9249** (below default 0.99) |
| Result | PASS with `assert_pcc: false` |
| Wall time | 4:22 |
| Log | `logs/iter10_verify_video_input.log` |

Follow-up to clear 0.99 (and to lift the 4-frame cap): shard the video tower across the mesh (currently replicated) and revisit KV replication.

**Note:** base (non-instruct) checkpoint + static replicated frames → degenerate caption; the bring-up goal was a finite, reproducible forward through the video path, not caption quality. [can be added in future]
